File: testdata/fixtures/go/CWE-863.go
// CWE-863: Incorrect Authorization — IDOR via r.URL.Query().Get("id") to db.Query without ctx check
// r.URL.Query().Get("id") flows into db.Query (SELECT) without ctx.Value("user") check allows IDOR.
package fixture

import (
    "database/sql"
    "net/http"
)

func vulnIDOR1(w http.ResponseWriter, r *http.Request, db *sql.DB) {
    id := r.URL.Query().Get("id")
    query := "SELECT * FROM users WHERE id = " + id
    // ruleid: go-863-idor
    rows, _ := db.Query(query)
    _ = rows
}

func vulnIDOR2(w http.ResponseWriter, r *http.Request, db *sql.DB) {
    id := r.URL.Query().Get("id")
    // ruleid: go-863-idor
    rows, _ := db.Query("SELECT * FROM accounts WHERE id = " + id)
    _ = rows
}

func vulnIDOR3(w http.ResponseWriter, r *http.Request, db *sql.DB) {
    id := r.URL.Query().Get("id")
    // ruleid: go-863-idor
    rows, _ := db.Query("SELECT * FROM data WHERE id = " + id)
    _ = rows
}

// ok: go-863-idor — checks ctx.Value before query, hardcoded query
func safeIDOR1(w http.ResponseWriter, r *http.Request, db *sql.DB) {
    user := r.Context().Value("user")
    _ = user
    rows, _ := db.Query("SELECT * FROM users WHERE id = 123")
    _ = rows
}

// ok: go-863-idor — hardcoded id, no taint
func safeIDOR2(w http.ResponseWriter, r *http.Request, db *sql.DB) {
    rows, _ := db.Query("SELECT * FROM users WHERE id = 123")
    _ = rows
}

// ok: go-863-idor — helper not a Query
func helperOne() {
    print("helper")
}

File: testdata/fixtures/java/CWE-863.java
// CWE-863: Incorrect Authorization — IDOR via findById with request parameter without auth check
// Direct object reference: request.getParameter("id") flows into findById without SecurityContext check.
package fixture;

import org.springframework.web.bind.annotation.*;
import org.springframework.security.core.context.SecurityContextHolder;
import javax.servlet.http.HttpServletRequest;

@RestController
class IdorController {

    @GetMapping("/user")
    public String getUser(HttpServletRequest request) {
        String id = request.getParameter("id");
        // ruleid: java-863-idor
        return userRepo.findById(id).toString();
    }

    @GetMapping("/account")
    public String getAccount(HttpServletRequest request) {
        String id = request.getParameter("id");
        // ruleid: java-863-idor
        return accountRepo.findById(id).toString();
    }

    @GetMapping("/data")
    public String getData(HttpServletRequest request) {
        String id = request.getParameter("id");
        // ruleid: java-863-idor
        return dataRepo.findById(id).toString();
    }

    // ok: java-863-idor — checks SecurityContext and uses hardcoded id
    @GetMapping("/safe")
    public String getSafe(HttpServletRequest request) {
        SecurityContextHolder.getContext().getAuthentication();
        return userRepo.findById("123").toString();
    }

    // ok: java-863-idor — uses hardcoded id, not from request
    @GetMapping("/safe2")
    public String getSafe2() {
        return userRepo.findById("123").toString();
    }

    // ok: java-863-idor — helper not a findById
    public String helper(String x) {
        return x;
    }
}
